File: switch.py
import boto3

def stop_auto_scaling_groups(region):
    ec2_client = boto3.client('ec2', region_name=region)
    autoscaling_client = boto3.client('autoscaling', region_name=region)

    # Get all Auto Scaling groups in the specified region
    response = autoscaling_client.describe_auto_scaling_groups()
    auto_scaling_groups = response['AutoScalingGroups']

    if not auto_scaling_groups:
        print("No Auto Scaling groups found in the specified region.")
        return

    for group in auto_scaling_groups:
        group_name = group['AutoScalingGroupName']
        
        # Get instance IDs from the Auto Scaling group
        instance_ids = [instance['InstanceId'] for instance in group['Instances']]
        ec2_client.terminate_instances(InstanceIds=instance_ids)

        autoscaling_client.update_auto_scaling_group(
            AutoScalingGroupName=group_name,
            DesiredCapacity=0,
            MaxSize=0,
            MinSize=0,
        )

        # Suspend scaling processes to prevent automatic instance launches
        autoscaling_client.suspend_processes(AutoScalingGroupName=group_name)

        print(f"All instances in Auto Scaling group {group_name} stopped and scaling processes suspended.")

def resume_auto_scaling_groups(region):
    autoscaling_client = boto3.client('autoscaling', region_name=region)

    # Get all Auto Scaling groups in the specified region
    response = autoscaling_client.describe_auto_scaling_groups()
    auto_scaling_groups = response['AutoScalingGroups']

    if not auto_scaling_groups:
        print("No Auto Scaling groups found in the specified region.")
        return

    for group in auto_scaling_groups:
        group_name = group['AutoScalingGroupName']

        autoscaling_client.update_auto_scaling_group(
            AutoScalingGroupName=group_name,
            DesiredCapacity=1,
            MaxSize=1,
            MinSize=1,
        )

        # Resume scaling processes
        autoscaling_client.resume_processes(AutoScalingGroupName=group_name)

        print(f"Scaling processes for Auto Scaling group {group_name} resumed.")

# Interactive CLI
def interactive_cli():
    region = input("Enter the AWS region: ")

    while True:
        action = input("Enter 'stop' to stop all instances, 'resume' to resume scaling processes, or 'q' to quit: ")

        if action == 'stop':
            stop_auto_scaling_groups(region)
        elif action == 'resume':
            resume_auto_scaling_groups(region)
        elif action == 'q':
            print("Exiting the program...")
            break
        else:
            print("Invalid input. Please try again.")

# Run
interactive_cli()

File: switch.go
package main

import (
	"fmt"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/aws/external"
	"github.com/aws/aws-sdk-go-v2/service/autoscaling"
	"github.com/aws/aws-sdk-go-v2/service/ec2"
)

func stopAutoScalingGroup(autoScalingGroupName string) {
	cfg, err := external.LoadDefaultAWSConfig()
	if err != nil {
		fmt.Println("Error loading AWS config:", err)
		return
	}

	ec2Svc := ec2.New(cfg)
	autoscalingSvc := autoscaling.New(cfg)

	// Get instance IDs from the Auto Scaling group
	asgInput := &autoscaling.DescribeAutoScalingGroupsInput{
		AutoScalingGroupNames: []string{autoScalingGroupName},
	}
	asgOutput, err := autoscalingSvc.DescribeAutoScalingGroups(asgInput)
	if err != nil {
		fmt.Println("Error:", err)
		return
	}

	var instanceIDs []string
	for _, instance := range asgOutput.AutoScalingGroups[0].Instances {
		instanceIDs = append(instanceIDs, aws.ToString(instance.InstanceId))
	}

	// Terminate instances
	terminateInput := &ec2.TerminateInstancesInput{
		InstanceIds: instanceIDs,
	}
	_, err = ec2Svc.TerminateInstancesRequest(terminateInput).Send()
	if err != nil {
		fmt.Println("Error:", err)
		return
	}

	// Set desired capacity to 0 to stop all instances
	updateInput := &autoscaling.UpdateAutoScalingGroupInput{
		AutoScalingGroupName: aws.String(autoScalingGroupName),
		DesiredCapacity:      aws.Int64(0),
		MaxSize:              aws.Int64(0),
		MinSize:              aws.Int64(0),
	}
	_, err = autoscalingSvc.UpdateAutoScalingGroupRequest(updateInput).Send()
	if err != nil {
		fmt.Println("Error:", err)
		return
	}

	// Suspend scaling processes to prevent automatic instance launches
	suspendInput := &autoscaling.SuspendProcessesInput{
		AutoScalingGroupName: aws.String(autoScalingGroupName),
	}
	_, err = autoscalingSvc.SuspendProcessesRequest(suspendInput).Send()
	if err != nil {
		fmt.Println("Error:", err)
		return
	}

	fmt.Println("All instances in Auto Scaling group", autoScalingGroupName, "stopped and scaling processes suspended.")
}

func resumeAutoScalingGroup(autoScalingGroupName string) {
	cfg, err := external.LoadDefaultAWSConfig()
	if err != nil {
		fmt.Println("Error loading AWS config:", err)
		return
	}

	autoscalingSvc := autoscaling.New(cfg)

	// Set max size to the total number of instances to allow instance launches
	updateInput := &autoscaling.UpdateAutoScalingGroupInput{
		AutoScalingGroupName: aws.String(autoScalingGroupName),
		DesiredCapacity:      aws.Int64(1),
		MaxSize:              aws.Int64(1),
		MinSize:              aws.Int64(1),
	}
	_, err = autoscalingSvc.UpdateAutoScalingGroupRequest(updateInput).Send()
	if err != nil {
		fmt.Println("Error:", err)
		return
	}

	// Resume scaling processes to allow automatic instance launches
	resumeInput := &autoscaling.ResumeProcessesInput{
		AutoScalingGroupName: aws.String(autoScalingGroupName),
	}
	_, err = autoscalingSvc.ResumeProcessesRequest(resumeInput).Send()
	if err != nil {
		fmt.Println("Error:", err)
		return
	}

	fmt.Println("Scaling processes for Auto Scaling group", autoScalingGroupName, "resumed.")
}


func interactiveCLI() {
	var action string
	for {
		fmt.Print("Enter 'stop' to stop all instances, 'resume' to resume scaling processes, or 'q' to quit: ")
		fmt.Scanln(&action)

		if action == "stop" {
			var autoScalingGroupName string
			fmt.Print("Enter the Auto Scaling group name: ")
			fmt.Scanln(&autoScalingGroupName)
			stopAutoScalingGroup(autoScalingGroupName)
		} else if action == "resume" {
			var autoScalingGroupName string
			fmt.Print("Enter the Auto Scaling group name: ")
			fmt.Scanln(&autoScalingGroupName)
			resumeAutoScalingGroup(autoScalingGroupName)
		} else if action == "q" {
			fmt.Println("Exiting the program...")
			break
		} else {
			fmt.Println("Invalid input. Please try again.")
		}
	}
}

func main() {
	interactiveCLI()
}
